### PR TITLE
iox-#1099 add test for error handler mock

### LINKED ISSUE #1099: Separate module specific errors from `iceoryx_hoofs`
## Brief feature description

#590 uncovered an illegal dependency from `iceoryx_hoofs` towards modules using it. While `iceoryx_hoofs` aims to be a library that [can be used independent](https://discourse.ros.org/uploads/short-url/cKUVQ4V8aUu7nB2D4IvldyozD64.pdf) from `iceoryx_posh` and the other modules, as of today it still contains all the error codes from all modules.

## Detailed information

* [x] Separate test specific code from the `ErrorHandler` and templatize `setTemporaryErrorHandler(..)`
* [x] Call `errorHandler(..)` only once in `iceoryx_hoofs` in `Expects` / `Ensures` 
* [x] Remove the 2nd argument of `errorHandler(..)` being `const std::function<void()>& errorCallBack IOX_MAYBE_UNUSED`
* [x] Create separate `error_handling.hpp` for each module using the `ErrorHandler`
   * [x] Rename this `error_handling.cpp` in `hoofs` to `error_handler.hpp`
* [x] Use `asStringLiteral` instead of `toString` from #1066 
* [x] Add explanation in error handling concept about what needs to be implemented for new module
* [ ] Add basic unit tests for `ErrorHandlerMock`

## References

* Comes from #590

### PR DESCRIPTION
**Pull request by mossmaurice**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] ~~Changelog updated [in the unreleased section][changelog] including API breaking changes~~
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

* As `ErrorHandlerMock` is an essential component in the tests, it shall be tested to ensure its correct behaviour

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1099

### CODE REVIEW

File: iceoryx_hoofs/test/mocktests/test_hoofs_mock.cpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
```
Comment (mossmaurice) [inline iceoryx_hoofs/test/mocktests/test_hoofs_mock.cpp:1]: This was copy & past'ed, hence the old copy right is kept here.

Comment (elfenpiff) [reply iceoryx_hoofs/test/mocktests/test_hoofs_mock.cpp:1]: Really? This looks like it was taken from the google test documentation ;)

But actually when you take a look at my code here from the racer peppi: https://gitlab.com/el.chris/racerpeppi/-/blob/master/unittest/unittest.cpp

I claim it was copied from there (line 23 - 27) and therefore I claim the copyright. Please add the following line:
```
// Copyright (C) All glory to the hypnotoad.
```

File: iceoryx_hoofs/test/mocktests/test_hoofs_mock.cpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "test.hpp"
+
+using namespace ::testing;
```
Comment (elfenpiff) [inline iceoryx_hoofs/test/mocktests/test_hoofs_mock.cpp:20]: Why write here using namespace and then still write `::testing::` in line 25?

File: iceoryx_hoofs/test/mocktests/test_error_handler_mock.cpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+TEST(ErrorHandlerMock_test, CallingErrorHandlerWithErrorOfKnownModuleAndDefaultLevelIsCaught)
+{
+    auto errorHandlerGuard =
+        iox::ErrorHandlerMock::setTemporaryErrorHandler<KnownError>([&](const auto error, const iox::ErrorLevel level) {
+            EXPECT_THAT(error, Eq(KnownError::TEST__FOOBAR));
+            EXPECT_THAT(level, Eq(iox::ErrorLevel::FATAL));
+        });
+    iox::errorHandler(KnownError::TEST__FOOBAR);
```
Comment (elfenpiff) [inline iceoryx_hoofs/test/mocktests/test_error_handler_mock.cpp:75]: I think when the error is not caught the test will still succeed.

Could you please set a variable in the callback and check after the error that the variable was set.

File: iceoryx_hoofs/test/mocktests/test_error_handler_mock.cpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+TEST(ErrorHandlerMock_test, CallingErrorHandlerWithErrorOfKnownModuleAndNonDefaultLevelIsCaught)
+{
+    auto errorHandlerGuard =
+        iox::ErrorHandlerMock::setTemporaryErrorHandler<KnownError>([&](const auto error, const iox::ErrorLevel level) {
+            EXPECT_THAT(error, Eq(KnownError::TEST__FOOBAR));
+            EXPECT_THAT(level, Eq(iox::ErrorLevel::MODERATE));
+        });
+    iox::errorHandler(KnownError::TEST__FOOBAR, iox::ErrorLevel::MODERATE);
```
Comment (elfenpiff) [inline iceoryx_hoofs/test/mocktests/test_error_handler_mock.cpp:85]: Same issue as in the previous case. Will succeed even when the error is not caught.

File: iceoryx_hoofs/test/mocktests/test_hoofs_mock.cpp (hunk 2 above)
Comment (mossmaurice) [reply iceoryx_hoofs/test/mocktests/test_hoofs_mock.cpp:20]: Good point, thanks! Removed the `using namespace`.

File: iceoryx_hoofs/test/mocktests/test_error_handler_mock.cpp (hunk 3 above)
Comment (mossmaurice) [reply iceoryx_hoofs/test/mocktests/test_error_handler_mock.cpp:75]: Good catch, thanks! Fixed.

**Review by elfenpiff — APPROVED** [review]

File: iceoryx_hoofs/test/CMakeLists.txt [hunk 5]
```diff
@@ -58,7 +59,18 @@ target_include_directories(${PROJECT_PREFIX}_moduletests PRIVATE .)
...
+    CXX_STANDARD ${ICEORYX_CXX_STANDARD}
+    POSITION_INDEPENDENT_CODE ON
+)
+
+# mock tests
+add_executable(${PROJECT_PREFIX}_mocktests ${MOCKTESTS_SRC})
+target_compile_options(${PROJECT_PREFIX}_mocktests PRIVATE ${TEST_CXX_FLAGS})
+target_include_directories(${PROJECT_PREFIX}_mocktests PRIVATE .)
+target_link_libraries(${PROJECT_PREFIX}_mocktests ${TEST_LINK_LIBS})
+set_target_properties(${PROJECT_PREFIX}_mocktests PROPERTIES
```
Comment (elBoberido) [inline iceoryx_hoofs/test/CMakeLists.txt:73]: This does not make sense if it does not run on the CI. Please either enable this for the CI or just add it to the moduletests

File: iceoryx_meta/tests.cmake [hunk 6]
```diff
@@ -30,6 +30,10 @@ if (BUILD_TEST)
     endif()
 
     ### create test targets without Timing tests
+
+    ### Only hoofs has mock tests
+    list(APPEND MOCKTEST_CMD COMMAND ./hoofs/test/hoofs_mocktests --gtest_filter=-*.TimingTest_* --gtest_output=xml:${CMAKE_BINARY_DIR}/testresults/hoofs_MockTestResults.xml)
```
Comment (elBoberido) [inline iceoryx_meta/tests.cmake:35]: @dkroenke FYI

**Review by elBoberido — APPROVED** [review]
Assuming @dkroenke is fine with this, I would also be fine with it

**Review by elfenpiff — APPROVED** [review]

**Pull request merged**